driver-sync/src/test/functional/com/mongodb/client/AbstractMongoCollectionTest.java

```diff
@@ -0,0 +1,318 @@
+    // properly to handle DBRef encoding and decoding
+    @Test
+    public void testDBRefEncodingAndDecoding() {
+        // given
+        Document doc = new Document("_id", 1)
+                               .append("ref", new DBRef("foo", 5))
+                               .append("refWithDB", new DBRef("db", "foo", 5));
+
+
+        // when
+        getCollection().insertOne(doc);
+
+        // then
+        assertEquals(doc, getCollection().find().first());
+    }
+
+    @Test
+    public void testJsonObjectEncodingAndDecoding() {
+        // given
+        MongoCollection<JsonObject> test = getCollection().withDocumentClass(JsonObject.class);
+        JsonObject json = new JsonObject("{\"_id\": {\"$oid\": \"5f5a5442306e56d34136dbcf\"}, \"hello\": 1}");
+
+        // when
+        test.insertOne(json);
+
+        // then
+        assertEquals(json, test.find().first());
+    }
+
+    @Test
+    public void testObjectIdToStringConversion() {
+        // given
+        CodecRegistry pojoCodecRegistry = fromRegistries(MongoClientSettings.getDefaultCodecRegistry(),
+                fromProviders(PojoCodecProvider.builder().automatic(true).build()));
+
+        MongoCollection<BsonRepresentationModel> test = getCollection()
+                .withDocumentClass(BsonRepresentationModel.class)
+                .withCodecRegistry(pojoCodecRegistry);
+        test.drop();
+
+        // when
+        test.insertOne(new BsonRepresentationModel(null, 1));
+
+        // then
+        BsonRepresentationModel first = test.find().first();
+        assertNotNull(first);
+        assertNotNull(first.getId());
+    }
+
+    @Test
+    public void testOperationsUseDocumentClassCodec() {
+
+        Codec<ShapeModelAbstract> shapeModelCodec = new Codec<ShapeModelAbstract>() {
+            private final CodecRegistry pojoCodecRegistry = fromRegistries(MongoClientSettings.getDefaultCodecRegistry(),
+                    fromProviders(PojoCodecProvider.builder().automatic(true).build()));
+
+            @Override
+            public ShapeModelAbstract decode(final BsonReader reader, final DecoderContext decoderContext) {
+                return pojoCodecRegistry.get(getEncoderClass()).decode(reader, decoderContext);
+            }
+
+            @Override
+            public void encode(final BsonWriter writer, final ShapeModelAbstract value, final EncoderContext encoderContext) {
+                pojoCodecRegistry.get(getEncoderClass()).encode(writer, value, encoderContext);
+            }
+
+            @Override
+            public Class<ShapeModelAbstract> getEncoderClass() {
+                return ShapeModelAbstract.class;
+            }
+        };
+        Codec<ShapeModelCircle> circleCodec = new Codec<ShapeModelCircle>() {
+
+            @Override
+            public void encode(final BsonWriter writer, final ShapeModelCircle value, final EncoderContext encoderContext) {
+                throw new UnsupportedOperationException("Not supported");
```

Comment:
IIUC we need these two custom Codec's here because, unlike bson-kotlinx, the PojoCodec will add a discriminator no matter which codec is used, the base class or the derived class.   And the test is relying on this method to throw in order to prove that the correct codec is used.  Is that correct?

If so, let's change `"Not Supported"` to something like `"If this method is called it means the document class codec was not properly used"`

Reply:
The premise is if the collection is `MongoCollection<Pet>` but the insert is of a `Cat` instance, it should use the codec for `Pet` not for `Cat`. So the test has two custom Codecs to prove that the correct one is picked.

With kotlinx serialization discriminators are declared via a [`PolymorphicSerializer`](https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-core/kotlinx.serialization/-polymorphic-serializer/index.html) and not on the data class itself.  So when looking up for a codec the `Pet` codec will use polymorphic serialization but the `Cat` codec won't and therefore be missing a discriminator.